index.js

```diff
@@ -16,23 +16,41 @@ export default async function run() {
     // Create a list of setup() promises from all the tools
     // Wait for all the setup() promises to resolve
     if (runAsync) {
-        core.info("Running setups in parallel")
-        const setups = Tool.all().map((tool) =>
-            tool.setup(core.getInput(tool.name)),
+        core.info("Running Tier 1 setups in parallel")
+        const tier1Setups = Tool.allTier1().map((tier1Tool) =>
+            tier1Tool.setup(core.getInput(tier1Tool.name)),
         )
-        return Promise.all(setups)
+        await Promise.all(tier1Setups)
+        core.info("Running Tier 2 setups in parallel")
+        const tier2Setups = Tool.allTier2().map((tier2Tool) =>
+            tier2Tool.setup(core.getInput(tier2Tool.name)),
+        )
+        return Promise.all(tier2Setups)
     } else {
         core.info("Running setups sequentially")
         let errs = []
         const errHandler = (err) => {
             core.error(`caught error in setup: ${err}`)
             errs.push(err)
         }
-        for (let tool of Tool.all()) {
+        // Force the first tier tools to be setup prior to the second tier as the second tier tools
+        // depend on the first tier tools already being available.
+        for (let tier1Tool of Tool.allTier1()) {
             try {
                 // For some reason this catch call isn't working the way I expect,
                 // but I can't figure it out, so we double down with try/catch
-                await tool.setup(core.getInput(tool.name)).catch(errHandler)
+                await tier1Tool
+                    .setup(core.getInput(tier1Tool.name))
+                    .catch(errHandler)
+            } catch (err) {
+                errHandler(err)
```

Comment:
If a tier1 tool fails, should we even try to install tier2 tools? Assuming that kotlin depends on java, it means that kotlin will probably fail too

Reply:
I don't know if you want to do tiers or better have tool dependencies and use a different model

Reply:
I suggested having a Tool depends on Tool kind of relationship, @shakefu suggested tiering.